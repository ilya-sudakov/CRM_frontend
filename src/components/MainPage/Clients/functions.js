export const clientsFormHeaderMenu = (name) => {
  return [
    { pageTitle: 'История работы', pageName: 'workHistory' },
    {
      pageTitle: `Данные ${name}а`,
      pageName: 'clientData',
    },
  ];
};
