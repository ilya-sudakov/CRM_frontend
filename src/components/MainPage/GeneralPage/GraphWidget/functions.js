export const getDateFromWeekdayIndex = (dayIndex, weekOffset) => {
  return new Date(
    new Date().setDate(
      new Date().getDate() +
        -7 * weekOffset -
        new Date(
          new Date().setDate(new Date().getDate() + -7 * weekOffset),
        ).getDay() +
        dayIndex,
    ),
  );
};

export const getWeekdaysListWithOffset = (offset = 0) => {
  const curDay = new Date(
    new Date().setDate(new Date().getDate() + -7 * offset),
  );
  let week = [];
  for (
    let i = 1;
    i <=
    (offset === 0 ? (new Date().getDay() === 0 ? 7 : new Date().getDay()) : 7);
    i++
  ) {
    let first = curDay.getDate() - curDay.getDay() + i;
    let day = new Date(curDay.setDate(first));
    week.push(day);
  }
  return week;
};
