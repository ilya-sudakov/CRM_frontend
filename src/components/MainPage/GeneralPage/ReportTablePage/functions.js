export const sortEmployees = (employees) => {
  return employees.sort((a, b) => {
    if (a.employee.lastName < b.employee.lastName) {
      return -1;
    }
    if (a.employee.lastName > b.employee.lastName) {
      return 1;
    }
    return 0;
  });
};

export const getMonthDates = (curDate) => {
  const date =
    new Date(curDate.getFullYear(), curDate.getMonth() + 1, 0).getDate() + 1;
  //Получаем массив с датами месяца
  let dates = [];
  for (let i = 1; i < date; i++) dates.push(i);

  return dates;
};
