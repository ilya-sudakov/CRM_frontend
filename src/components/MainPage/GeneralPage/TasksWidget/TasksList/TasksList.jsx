import PlaceholderLoading from '../../../../../utils/TableView/PlaceholderLoading/PlaceholderLoading.jsx';
import {
  formatDateString,
  formatDateStringNoYear,
  dateDiffInDays,
} from '../../../../../utils/functions.jsx';
import { conditions } from '../objects.js';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';

const TasksList = ({
  tasks = [],
  isLoading = false,
  controlDates = {},
  userHasAccess = () => {},
}) => {
  return (
    <div className="tasks-widget__list">
      {isLoading ? (
        <PlaceholderLoading
          itemClassName="list__item list__item--placeholder"
          minHeight="2rem"
          items={3}
        />
      ) : (
        Object.entries(controlDates).map((date) => {
          const isExpired = new Date(date[0]) < new Date();
          return (
            <ListWrapper
              isExpired={isExpired}
              date={date}
              tasks={tasks}
              userHasAccess={userHasAccess}
            />
          );
        })
      )}
    </div>
  );
};

export default TasksList;

TasksList.propTypes = {
  tasks: PropTypes.array,
  isLoading: PropTypes.bool,
  controlDates: PropTypes.object,
  userHasAccess: PropTypes.func,
};

const ListWrapper = ({
  isExpired = false,
  date = [new Date()],
  tasks = [],
  userHasAccess,
}) => {
  return (
    <div
      className={`tasks-widget__date-wrapper ${
        isExpired ? 'tasks-widget__date-wrapper--expired' : ''
      }`}
    >
      <div
        className={`tasks-widget__date ${
          isExpired ? 'tasks-widget__date--expired' : ''
        }`}
      >{`до ${formatDateStringNoYear(date[0])} ${
        isExpired
          ? `- опоздание ${dateDiffInDays(new Date(date[0]), new Date())} дн.`
          : ''
      }`}</div>
      {tasks
        .filter(
          (task) =>
            formatDateString(task.dateControl) === formatDateString(date[0]),
        )
        .map((task) => (
          <ListItem task={task} userHasAccess={userHasAccess} />
        ))}
    </div>
  );
};

ListWrapper.propTypes = {
  tasks: PropTypes.array,
  index: PropTypes.number,
  isExpired: PropTypes.bool,
  date: PropTypes.array,
  userHasAccess: PropTypes.func,
};

const ListItem = ({
  task = {
    id: 1,
    description: '',
    condition: '',
    status: '',
    dateCreated: new Date(),
    dateControl: new Date(),
  },
  userHasAccess,
}) => {
  const history = useHistory();

  return (
    <div
      className={`list__item list__item--${conditions[task.condition]}`}
      key={task.id}
    >
      <div
        className="list-item__general-info"
        onClick={() =>
          history.push(`/dispatcher/general-tasks/edit/${task.id}`)
        }
      >
        <span className="list-item__description">{task.description}</span>
        <span className="list-item__conditions">
          <span className="condition condition--status">{task.condition}</span>
          <span
            className={`condition condition--date ${
              !userHasAccess(['ROLE_ADMIN']) &&
              (task.status === '' || task.status === null)
                ? 'condition--hidden'
                : ''
            }`}
          >{`от ${formatDateStringNoYear(task.dateCreated)}`}</span>
          <span className="condition condition--description">
            {userHasAccess(['ROLE_ADMIN']) ? task.responsible : task.status}
          </span>
        </span>
      </div>
    </div>
  );
};

ListItem.propTypes = {
  task: PropTypes.object,
  userHasAccess: PropTypes.func,
};
