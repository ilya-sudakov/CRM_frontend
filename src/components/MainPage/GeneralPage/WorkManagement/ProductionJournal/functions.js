import {
  formatDateStringNoYear,
  getEmployeeNameText,
} from "../../../../../utils/functions.jsx";
import {
  createPDF,
  defaultStylesPDF,
} from "../../../../../utils/pdfFunctions.js";
import { sortByField } from "../../../../../utils/sorting/sorting.js";
import { defaultJournalWorkshops } from "./objects.js";

export const updateData = (worksList, selectedWork, newData) => {
  const temp = worksList;
  temp.splice(worksList.indexOf(selectedWork), 1, newData);
  return temp;
};

const getProductListText = (products) => {
  return products
    .map((product) => `${product.name} - ${product.quantity} шт\n`)
    .join("");
};

const getDaysWorkText = (workItem) => {
  let workText = workItem.works.map((work) => {
    const productText = getProductListText(work.product);
    const draftText = getProductListText(work.draft);
    return [
      `${work.workName} - ${work.hours} ч\n`,
      { text: productText, fontSize: 10, color: "#666" },
      { text: draftText, fontSize: 10, color: "#666" },
      work.comment && {
        text: `Комментарий: ${work.comment}`,
        fontSize: 10,
        color: "#999",
        italics: true,
      },
    ];
  });
  return workText;
};

const sortEmployees = (employees, workshop) => {
  const filteredEmployees = sortByField(
    employees.filter(
      (employee) =>
        employee.workshop === workshop.name &&
        employee?.relevance !== "Уволен" &&
        employee?.workshop !== "Уволенные"
    ),
    {
      fieldName: "lastName",
      direction: "asc",
    }
  );
  return filteredEmployees;
};

const getWorkshopList = (
  employees,
  workshop,
  curDate,
  todaysWork,
  yesterdaysWork
) => {
  const prevDay = new Date(new Date(curDate).setDate(curDate.getDate() - 1));
  const filteredEmployees = sortEmployees(employees, workshop);
  if (filteredEmployees.length === 0) return null;
  const listItems = filteredEmployees.map((employee) => [
    { text: getEmployeeNameText(employee), style: "regularText", fontSize: 11 },
    {
      stack: getDaysWorkText(yesterdaysWork[workshop.engName][employee.id]),
      style: "regularText",
      fontSize: 11,
    },
    {
      stack: getDaysWorkText(todaysWork[workshop.engName][employee.id]),
      style: "regularText",
      fontSize: 11,
    },
  ]);
  const list = [
    {
      text: `${workshop.name}\n`,
      alignment: "left",
      style: "subheader",
      margin: [0, 15, 0, 5],
    },
    {
      table: {
        widths: [150, "*", "*"],
        body: [
          [
            { text: "ФИО сотрудника", style: "tableHeader" },
            {
              text: formatDateStringNoYear(prevDay),
              style: "tableHeader",
            },
            { text: formatDateStringNoYear(curDate), style: "tableHeader" },
          ],
          ...listItems,
        ],
      },
    },
  ];
  return list;
};

export const createWorkListPDF = (
  employees,
  todaysWork,
  yesterdaysWork,
  curDate
) => { 
  const list = [];
  Object.values(defaultJournalWorkshops).map((workshop) => {
    list.push(
      getWorkshopList(employees, workshop, curDate, todaysWork, yesterdaysWork)
    );
  });
  const data = {
    info: {
      title: "Дневник производства",
    },
    content: [
      {
        text: "Дневник производства\n",
        alignment: "center",
        style: "title",
        margin: [0, 0, 0, 10],
      },
      ...list,
    ],
    styles: defaultStylesPDF,
  };
  createPDF(data);
};
