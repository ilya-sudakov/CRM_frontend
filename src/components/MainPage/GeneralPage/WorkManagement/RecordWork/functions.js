import {
  addDraftToRecordedWork,
  addProductToRecordedWork,
  addRecordedWork,
  deleteDraftFromRecordedWork,
  deleteProductFromRecordedWork,
  deleteRecordedWork,
  editRecordedWork,
} from 'Utils/RequestsAPI/WorkManaging/WorkControl.jsx';

export const submitWorkData = async (
  worktimeInputs,
  date,
  employee,
  setIsLoading,
  inputs,
) => {
  setIsLoading(true);
  return await Promise.all(
    worktimeInputs.works.map(async (item) => {
      console.log(item);
      const temp = Object.assign({
        day: date.getDate(),
        month: date.getMonth() + 1,
        year: date.getFullYear(),
        employeeId: employee.id,
        comment: item.comment,
        workListId: item.workId,
        hours: item.hours,
      });

      //Удаление элементов
      await Promise.all(
        worktimeInputs.originalWorks.map((originalWork) => {
          const item = worktimeInputs.works.find(
            (workItem) => workItem.id === originalWork.id,
          );
          if (originalWork.id && item === undefined) {
            console.log('deleting element', worktimeInputs);
            return Promise.all(
              originalWork.product.map((product) => {
                return deleteProductFromRecordedWork(
                  originalWork.id,
                  product.product.id,
                );
              }),
            )
              .then(() => {
                return Promise.all(
                  originalWork.draft.map((draft) => {
                    return deleteDraftFromRecordedWork(
                      originalWork.id,
                      draft.partId,
                      draft.partType,
                    );
                  }),
                );
              })
              .then(() => deleteRecordedWork(originalWork.id));
          }
        }),
      );

      if (item.isOld) {
        //если часы/комментарий не совпадают, то редактируем
        const originalItem = worktimeInputs.originalWorks.find(
          (originalWork) => item.id === originalWork.id,
        );

        if (
          originalItem &&
          (item.hours !== originalItem.hours ||
            item.comment !== originalItem.comment ||
            item.workName !== originalItem.workName)
        ) {
          await editRecordedWork(temp, item.id);
        }

        //Продукция
        //delete removed products
        Promise.all(
          originalItem.product.map((originalProduct) => {
            if (
              item.product.find(
                (product) => product.id === originalProduct.id,
              ) === undefined
            ) {
              console.log('delete product', originalProduct);
              return deleteProductFromRecordedWork(item.id, originalProduct.id);
            }
          }),
        );

        //add new products
        Promise.all(
          item.product.map((product) => {
            if (
              originalItem.product.find(
                (originalProduct) => product.id === originalProduct.id,
              ) === undefined
            ) {
              console.log('add product', product);
              return addProductToRecordedWork(
                item.id,
                product.id,
                product.quantity,
                {
                  name: product.name,
                },
              );
            }
          }),
        );

        //update edited products
        Promise.all(
          originalItem.product.map((originalProduct) => {
            const product = item.product.find(
              (product) => product.id === originalProduct.id,
            );
            if (
              product !== undefined &&
              originalProduct.quantity !== Number.parseFloat(product.quantity)
            ) {
              console.log('edit product', product, originalProduct);
              return deleteProductFromRecordedWork(
                item.id,
                originalProduct.id,
              ).then(() =>
                addProductToRecordedWork(
                  item.id,
                  product.id,
                  product.quantity,
                  {
                    name: product.name,
                  },
                ),
              );
            }
          }),
        );

        //Чертежи
        //delete removed drafts
        Promise.all(
          originalItem.draft.map((originalDraft) => {
            if (
              item.draft.find((draft) => draft.id === originalDraft.id) ===
              undefined
            ) {
              console.log('delete draft', originalDraft, originalItem);
              return deleteDraftFromRecordedWork(
                item.id,
                originalDraft.partId,
                originalDraft.partType,
              );
            }
          }),
        );

        //add new drafts
        Promise.all(
          item.draft.map((draft) => {
            if (
              originalItem.draft.find(
                (originalDraft) => draft.id === originalDraft.id,
              ) === undefined
            ) {
              console.log('add draft', draft);
              return addDraftToRecordedWork(
                item.id,
                draft.partId,
                draft.type,
                draft.quantity,
                draft.name,
              );
            }
          }),
        );

        //update edited drafts
        Promise.all(
          originalItem.draft.map((originalDraft) => {
            const draft = item.draft.find(
              (draft) => draft.id === originalDraft.id,
            );
            console.log('edit draft opportunity', originalDraft, draft);
            if (
              draft !== undefined &&
              originalDraft.quantity !== Number.parseFloat(draft.quantity)
            ) {
              console.log('edit draft success', draft);
              return deleteDraftFromRecordedWork(
                item.id,
                originalDraft.partId,
                originalDraft.partType ??
                  item.draft.find((item) => draft.partId === item.id)?.type,
              ).then(() =>
                addDraftToRecordedWork(
                  item.id,
                  draft.partId,
                  draft.partType,
                  draft.quantity,
                  draft.name,
                ),
              );
            }
          }),
        );

        inputs && inputs.updateSelectedDaysWork(item);
      }

      //if item is new, then just add it
      if (!item.isOld && item.workId !== null && item.isOld !== undefined) {
        console.log('adding item', item);
        return addRecordedWork(temp)
          .then((res) => res.json())
          .then((res) => {
            Promise.all(
              item.product.map((product) => {
                return addProductToRecordedWork(
                  res.id,
                  product.id,
                  product.quantity,
                  {
                    name: product.name,
                  },
                );
              }),
            );
            return res.id;
          })
          .then((id) => {
            return Promise.all(
              item.draft.map((draft) => {
                return addDraftToRecordedWork(
                  id,
                  draft.partId,
                  draft.type,
                  draft.quantity,
                  draft.name,
                );
              }),
            );
          })
          .then(() => inputs && inputs.addSelectedDaysWork(item))
          .catch((error) => {
            alert('Ошибка при добавлении записи');
            setIsLoading(false);
            console.log(error);
          });
      }
    }),
  ).then(() => {
    setIsLoading(false);
  });
};
