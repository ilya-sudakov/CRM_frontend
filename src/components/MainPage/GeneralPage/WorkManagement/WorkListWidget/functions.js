export const sortEmployeesObject = (employees) => {
  return employees.sort((a, b) => {
    if (a[1].lastName < b[1].lastName) {
      return -1;
    }
    if (a[1].lastName > b[1].lastName) {
      return 1;
    }
    return 0;
  });
};

export const filterEmployeesObject = (employees, workshop) => {
  return employees.filter((employee) => {
    const item = employee[1];
    if (item.workshop === workshop.name) return true;
  });
};
