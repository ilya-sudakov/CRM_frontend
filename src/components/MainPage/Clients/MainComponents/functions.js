import { exportClientsEmailsCSV } from 'Utils/xlsxFunctions.js';
import { getClients } from 'Utils/RequestsAPI/Clients.jsx';

export const sortClients = (clients, searchQuery, sortOrder) => {
  return clients.sort((a, b) => {
    if (searchQuery !== '') {
      const first =
          sortOrder.curSort === 'nextDateContact'
            ? new Date(a[sortOrder.curSort])
            : a[sortOrder.curSort],
        second =
          sortOrder.curSort === 'nextDateContact'
            ? new Date(b[sortOrder.curSort])
            : b[sortOrder.curSort];
      if (first < second) {
        return sortOrder[sortOrder.curSort] === 'desc' ? 1 : -1;
      }
      if (first > second) {
        return sortOrder[sortOrder.curSort] === 'desc' ? -1 : 1;
      }
      return 0;
    }
  });
};

export const getEmailsExcel = () => {
  let totalClients = 1;
  let clients = [];
  return getClients(1)
    .then((res) => res.json())
    .then((res) => {
      totalClients = res.totalElements;
      return getClients(totalClients);
    })
    .then((res) => res.json())
    .then((res) => {
      clients = res.content;
      console.log(clients);
      exportClientsEmailsCSV(clients);
    })
    .catch((error) => {
      console.log(error);
    });
};
