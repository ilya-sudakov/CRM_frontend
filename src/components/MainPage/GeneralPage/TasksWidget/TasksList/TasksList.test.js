import TasksList from './TasksList.jsx';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {
  dateDiffInDays,
  formatDateStringNoYear,
} from '../../../../../utils/functions.jsx';
import { renderWithRouter } from '../../../../../utils/testing/functions.js';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('TasksList component', () => {
  afterEach(cleanup);

  it('matches snapshot', () => {
    const { asFragment } = render(<TasksList />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders w/o props', () => {
    render(<TasksList />);
  });

  it('renders loading', () => {
    render(<TasksList isLoading={true} />);
  });

  it('renders tasks list', async () => {
    const tasks = [
      {
        id: 1,
        description: 'asd',
        condition: '',
        status: '',
        dateControl: new Date(2019, 10, 12),
        dateCreated: new Date(2020, 10, 12),
      },
      {
        id: 2,
        description: '123',
        condition: '',
        status: null,
        dateControl: new Date(2019, 10, 12),
        dateCreated: new Date(2020, 10, 12),
      },
      {
        id: 3,
        description: 'zzz',
        condition: '',
        status: 'cvnbcvb',
        dateControl: new Date(2200, 10, 12),
        dateCreated: new Date(2120, 10, 12),
      },
    ];

    const controlDates = {
      [new Date(2019, 10, 12)]: [...tasks],
      [new Date(2200, 10, 12)]: [tasks[2]],
    };

    renderWithRouter(
      <TasksList
        userHasAccess={() => true}
        controlDates={controlDates}
        tasks={tasks}
      />,
    );
    expect(await screen.findByText('asd')).toBeInTheDocument();
    expect(await screen.findByText('123')).toBeInTheDocument();
    fireEvent.click(
      document.getElementsByClassName('list-item__general-info')[0],
    );
    expect(mockHistoryPush).toHaveBeenCalledWith(
      '/dispatcher/general-tasks/edit/1',
    );
  });

  it('renders expired task', async () => {
    const tasks = [
      {
        id: 1,
        description: 'asd',
        condition: '',
        status: '',
        dateControl: new Date(2019, 10, 12),
        dateCreated: new Date(2020, 10, 12),
      },
    ];

    const controlDates = { [new Date(2019, 10, 12)]: [...tasks] };

    renderWithRouter(
      <TasksList
        userHasAccess={() => true}
        controlDates={controlDates}
        tasks={tasks}
      />,
    );
    expect(
      await screen.findByText(
        `до ${formatDateStringNoYear(
          new Date(2019, 10, 12),
        )} - опоздание ${dateDiffInDays(
          new Date(2019, 10, 12),
          new Date(),
        )} дн.`,
      ),
    ).toBeInTheDocument();
  });
});
