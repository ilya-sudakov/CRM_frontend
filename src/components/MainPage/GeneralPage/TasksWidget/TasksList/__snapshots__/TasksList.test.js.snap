// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="tasks-widget__list"
  />
</DocumentFragment>
`;
